src/core/queue/BaseQueue.ts

```diff
@@ -0,0 +1,111 @@
+/* eslint-disable no-shadow */
+export enum QueueJobPriority {
+  HIGH = 1,
+  MEDIUM = 2,
+  LOW = 3
+}
+
+export type IQueueJob = {
+  task: () => Promise<any>;
+  label?: string;
+  callback?: (data) => void;
+};
+
+export type IPriorityQueueJob = IQueueJob & {
+  priority: QueueJobPriority.HIGH | QueueJobPriority.MEDIUM | QueueJobPriority.LOW;
+  createdAt?: number;
+};
+
+export type IQueue = {
+  addJob: (job: IQueueJob) => void;
+  addPriorityJob: (job: IPriorityQueueJob) => void;
+  processJobs: () => Promise<void>;
+  getRemainingJobsCount: () => number;
+};
+
+class BaseQueue {
+  static instance: BaseQueue;
+
+  jobs: IQueueJob[] = [];
+
+  highPriorityJobs: IPriorityQueueJob[] = [];
+
+  isExecutingJob = false;
+
+  static getInstance(): BaseQueue {
+    if (!this.instance) {
+      return new BaseQueue();
+    }
+
+    return this.instance;
+  }
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The singleton pattern implementation here is incorrect. The `getInstance` method should assign the newly created instance to `this.instance` if it doesn't exist, so that subsequent calls return the same instance.

```diff
  static getInstance(): BaseQueue {
    if (!this.instance) {
-     return new BaseQueue();
+     this.instance = new BaseQueue();
    }

    return this.instance;
  }
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->